Convert instructor email messages to ACE

OmarIthawi opened the pull request:
Convert instructor email messages to ACE:

**Caveat:** Due to a shortcoming in ACE, when the username is missing (e.g. for messages that are for emails to unregistered users), the file will be hidden in the in your File Manager `openedx/src/ace_messages/.20180628-195623.html`. Use the command `$ ls -lah` to find it.

### TODO

 - [x] Make Jenkins happy
 - [x] Add tests to the HTML message
 - [x] Support CCX messages as well
 - [x] Fix the ugly `course_name=display_name|default:course.display_name_with_default` and abstract it somewhere else
 - [x] Test all of the HTML messages manually:
   * [x] AccountCreationAndEnrollment
   * [x] AddBetaTester
   * [x] AllowedEnroll
   * [x] AllowedUnenroll
   * [x] EnrollEnrolled
   * [x] EnrolledUnenroll
   * [x] RemoveBetaTester
 - [x] Move `EmailTemplateTagMixin` into the `mixins.py` file
 - [x] **Question to Reviewer:** What about `test_site` and `stanford-style`? (I've removed them)
 - [x] @nedbat (or anyone from edX) to check with the white-label team before merging this PR

### Screenshots?
Screenshots are available in the comments below.

edx-webhook commented:
Thanks for the pull request, @OmarIthawi! I've created [OSPR-2490](https://openedx.atlassian.net/browse/OSPR-2490) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

OmarIthawi commented:
@nasthagiri FYI. No need to review yet. But just in case someone at edX wanted to do work on the conversion, this email message shouldn't need a conversion.

mduboseedx commented:
@OmarIthawi Thanks for the contribution. Feel free to ping me if you believe this will be ready before Friday.

OmarIthawi commented:
@mduboseedx Except for the remaining TODO's :point_up: I've completed a working copy of the PR.

Would you like to take a look?

OmarIthawi commented:
jenkins run python

OmarIthawi commented:
jenkins run bokchoy

OmarIthawi commented:
@mduboseedx The PR is ready for review. There are some obvious niptucks that I'm going to do once the review is done.

marcotuts commented:
I have no top level concerns around the product review, though I'm not familiar with which GUI you are referring to. Where is this being exposed normally when the username is included? 

Otherwise, good to move to engineering prioritization.

OmarIthawi commented:
Thanks @marcotuts! 

When a learner is being enrolled in a course by the staff this email is being sent to them:

```
# enroll_email_enrolledmessage.txt
...
You have been enrolled in {course_name} at {site_name} by a member
....
```

This PR turns into an HTML message using the relatively recent addition to Open edX https://github.com/edx/edx-ace

A similar work has been done for Password Reset email and now it's on production:

 - https://github.com/edx/edx-platform/pull/16545

I'm planning to open more of those PRs for other email messages. As far as I know there's no need to block engineering review for those PRs by product review, instead we should make it as a sanity checking last step before merging.

What do you think?

OmarIthawi commented:
jenkins run bokchoy

MHaton commented on lms/templates/instructor/edx_ace/enrollenrolled/email/body.html:
```diff
@@ -0,0 +1,30 @@
+                <br />
+            </p>
+
+            {% trans "Acces the Course Materials Now" as course_cta_text %}
```
Typo - "Acces" -> "Access :)

OmarIthawi replied:
Thanks @MHaton :)

OmarIthawi commented:
jenkins run bokchoy

mduboseedx commented:
@marcotuts Is it ok to move ahead with engineering review for @OmarIthawi's upcoming PRs for other email messages?

nedbat commented on lms/djangoapps/ccx/tests/utils.py:
```diff
@@ -8,14 +8,15 @@
 
 from lms.djangoapps.ccx.overrides import override_field_for_ccx
 from lms.djangoapps.ccx.tests.factories import CcxFactory
+from openedx.core.djangoapps.ace_common.tests.test_templatetags import EmailTemplateTagMixin
```
We should move EmailTemplateTagMixin to mixins.py in that directory, rather than importing from a test file.

OmarIthawi commented:
@nedbat I've regarding [you question about the caveat](https://github.com/edx/edx-platform/pull/18475#issuecomment-419136721) updated it to reflect what I meant. It's relevant when you test things in devstack, it's a shortcoming in ACE.

nedbat commented:
@OmarIthawi Thanks, now I understand. For debugging, Django can write emails to files instead of sending them, but the file name chosen in this case starts with a dot, and so might be hidden.

nedbat commented:
@OmarIthawi What else are you planning to do before merging this?

OmarIthawi commented:
Thanks @nedbat! Just saw your message. 

I'm planning to get some help in reviewing the HTML and Text content for the emails for look and feel and possible typos.

It takes up to a couple of hours to go through them, and as an author, I won't be the best person to review my own work, for the same reasons we get code reviews here.

OmarIthawi commented:
> @OmarIthawi What else are you planning to do before merging this?

@nedbat One more thing, I think that the edX white-label team should be informed about the change before this is merged.

OmarIthawi commented:
Status: I'm doing final check on this branch manually, I'll post both text and HTML emails in comments here.

MHaton commented on conf/locale/eo/LC_MESSAGES/django.po:
```diff
@@ -11211,6 +11211,413 @@ msgstr ""
+#, python-format
+msgid ""
+"To finish your registration, please visit %(registration_url)s and fill out "
+"the registration form making sure to use %(email_address)s in the E-mail "
```
E-mail -> Email

MHaton commented on conf/locale/eo/LC_MESSAGES/django.po:
```diff
@@ -11211,6 +11211,413 @@ msgstr ""
+
+#: lms/templates/instructor/edx_ace/allowedenroll/email/body.html
+msgid ""
+"Once you have registered and activated your account, you an access the "
```
you an access -> you will be able to access

MHaton commented on conf/locale/eo/LC_MESSAGES/django.po:
```diff
@@ -11211,6 +11211,413 @@ msgstr ""
+#: lms/templates/instructor/edx_ace/allowedunenroll/email/subject.txt
+#: lms/templates/instructor/edx_ace/enrolledunenroll/email/subject.txt
+#, python-format
+msgid "You have been un-enrolled from %(course_name)s"
```
un-enrolled -> unenrolled

MHaton commented on conf/locale/eo/LC_MESSAGES/django.po:
```diff
@@ -11211,6 +11211,413 @@ msgstr ""
+#: lms/templates/instructor/edx_ace/allowedunenroll/email/body.txt
+#, python-format
+msgid ""
+"You have been un-enrolled from course %(course_name)s by a member of the "
```
un-enrolled -> unenrolled (basically just replace it everywhere it happens)
...technically the term should be disenrolled, but that just sounds weird and everyone uses unenrolled so meh.

MHaton commented on conf/locale/eo/LC_MESSAGES/django.po:
```diff
@@ -11211,6 +11211,413 @@ msgstr ""
+"däshßöärd. Ⱡ'σяєм ιρѕυ#"
+
+#: lms/templates/instructor/edx_ace/enrollenrolled/email/body.html
+msgid "Acces the Course Materials Now"
```
Acces -> Access

MHaton commented on conf/locale/eo/LC_MESSAGES/django.po:
```diff
@@ -29584,27 +29747,30 @@ msgstr "À fïlé wäs délétéd: %s Ⱡ'σяєм ιρѕυм ∂σłσя ѕιт
-"    The file may be referenced on other articles. Deleting it means that they will loose their references to this file. The following articles reference this file:\n"
-"    "
+"The file may be referenced on other articles. Deleting it means that they "
+"will loose their references to this file. The following articles reference "
```
loose -> lose

MHaton commented on conf/locale/eo/LC_MESSAGES/django.po:
```diff
@@ -29584,27 +29747,30 @@ msgstr "À fïlé wäs délétéd: %s Ⱡ'σяєм ιρѕυм ∂σłσя ѕιт
-"\n"
-"    The file may be referenced on other articles. Deleting it means that they will loose their references to this file. The following articles reference this file:\n"
-"    "
+"The file may be referenced on other articles. Deleting it means that they "
```
the -> this

MHaton commented on conf/locale/eo/LC_MESSAGES/django.po:
```diff
@@ -29645,13 +29811,21 @@ msgstr "Ûsé thïs! Ⱡ'σяєм ιρѕυм ∂σł#"
+"Replacing an attachment means adding a new file that will be used in its "
+"place. All references to the file will be replaced by the one you upload and"
+" the file will be downloaded as <strong>%(filename)s</strong>. Please note "
+"that this attachment is in use on other articles, you may distort contents. "
```
"that if this attachment is in use on other articles, you may affect them in an unanticipated way."

MHaton commented on conf/locale/eo/LC_MESSAGES/django.po:
```diff
@@ -29645,13 +29811,21 @@ msgstr "Ûsé thïs! Ⱡ'σяєм ιρѕυм ∂σł#"
+" the file will be downloaded as <strong>%(filename)s</strong>. Please note "
+"that this attachment is in use on other articles, you may distort contents. "
+"However, do not hestitate to take advantage of this and make replacements "
+"for the listed articles where necessary. This way of working is more "
```
"for the listed articles where necessary, as it is typically more efficient to work in this way."

MHaton commented on lms/templates/instructor/edx_ace/accountcreationandenrollment/email/body.html:
```diff
@@ -0,0 +1,41 @@
+            </p>
+
+            <p style="color: rgba(0,0,0,.75);">
+                {% blocktrans %}Sincerely yours, The {{ course_name }} Team{% endblocktrans %}
```
Sincerely yours -> Your sincerely

MHaton commented on lms/templates/instructor/edx_ace/accountcreationandenrollment/email/body.html:
```diff
@@ -0,0 +1,41 @@
+                <br />
+                {% blocktrans %}email: {{ email_address }}{% endblocktrans %}
+                <br />
+                {% blocktrans %}password: {{ password }}{% endblocktrans %}
```
Really? there's a template that sends passwords in plaintext emails? 😱

OmarIthawi replied:
Yeah, I guess :slightly_frowning_face:

MHaton commented on lms/templates/instructor/edx_ace/accountcreationandenrollment/email/body.txt:
```diff
@@ -0,0 +1,11 @@
+
+{% blocktrans %}It is recommended that you change your password.{% endblocktrans %}
+
+{% blocktrans %}Sincerely yours, The {{ course_name }} Team{% endblocktrans %}{% endautoescape %}
```
Sincerely yours -> Yours sincerely

MHaton commented on lms/templates/instructor/edx_ace/allowedenroll/email/body.html:
```diff
@@ -0,0 +1,52 @@
+                {% if auto_enroll %}
+                    {% blocktrans %}Once you have registered and activated your account, you will see {{ course_name }} listed on your dashboard.{% endblocktrans %}
+                {% elif course_about_url is not None %}
+                    {% blocktrans %}Once you have registered and activated your account, you an access the course:{% endblocktrans %}
```
you an access -> you will be able to access

MHaton commented on lms/templates/instructor/edx_ace/allowedenroll/email/body.txt:
```diff
@@ -0,0 +1,14 @@
+{% elif course_about_url is not None %}
+{% blocktrans %}To access the course visit {{ course_about_url }} and register for the course.{% endblocktrans %}
+{% endif %}
+{% else %}{% blocktrans %}To finish your registration, please visit {{ registration_url }} and fill out the registration form making sure to use {{ email_address }} in the E-mail field.{% endblocktrans %}
```
E-mail -> Email

MHaton commented on lms/templates/instructor/edx_ace/allowedunenroll/email/body.html:
```diff
@@ -0,0 +1,25 @@
+            </h1>
+
+            <p style="color: rgba(0,0,0,.75);">
+                {% blocktrans %}You have been un-enrolled from course {{ course_name }} by a member of the course staff. Please disregard the invitation previously sent.{% endblocktrans %}
```
un-enrolled -> unenrolled
from course -> from the course

Honestly not sure about "previously sent" but it's probably fine.

MHaton commented on lms/templates/instructor/edx_ace/allowedunenroll/email/body.html:
```diff
@@ -0,0 +1,25 @@
+    <tr>
+        <td>
+            <h1>
+                {% blocktrans %}You have been un-enrolled from {{ course_name }}{% endblocktrans %}
```
un-enrolled -> unenrolled

MHaton commented on lms/templates/instructor/edx_ace/allowedunenroll/email/body.txt:
```diff
@@ -0,0 +1,6 @@
+{% load i18n %}{% autoescape off %}{% blocktrans %}Dear Student,{% endblocktrans %}
+
+{% blocktrans %}You have been un-enrolled from course {{ course_name }} by a member of the course staff. Please disregard the invitation previously sent.{% endblocktrans %}
```
un-enrolled -> unenrolled

MHaton commented on lms/templates/instructor/edx_ace/allowedunenroll/email/subject.txt:
```diff
@@ -0,0 +1,4 @@
+{% load i18n %}
+{% autoescape off %}
+{% blocktrans %}You have been un-enrolled from {{ course_name }}{% endblocktrans %}
```
un-enrolled -> unenrolled

MHaton commented on lms/templates/instructor/edx_ace/enrolledunenroll/email/body.html:
```diff
@@ -0,0 +1,32 @@
+            </h1>
+
+            <p style="color: rgba(0,0,0,.75);">
+                {% blocktrans %}You have been un-enrolled in {{ course_name }} at {{ site_name }} by a member of the course staff. The course will no longer appear on your {{ site_name }} dashboard.{% endblocktrans %}
```
un-enrolled -> unenrolled

The course -> This course

MHaton commented on lms/templates/instructor/edx_ace/enrolledunenroll/email/body.html:
```diff
@@ -0,0 +1,32 @@
+        <td>
+            <h1>
+                {% blocktrans %}
+                    You have been un-enrolled from {{ course_name }}
```
un-enrolled -> unenrolled

MHaton commented on lms/templates/instructor/edx_ace/enrolledunenroll/email/body.txt:
```diff
@@ -0,0 +1,8 @@
+{% load i18n %}{% autoescape off %}{% blocktrans %}Dear {{ full_name }}{% endblocktrans %}
+
+{% blocktrans %}You have been un-enrolled in {{ course_name }} at {{ site_name }} by a member of the course staff. The course will no longer appear on your {{ site_name }} dashboard.{% endblocktrans %}
```
un-enrolled -> unenrolled
The course -> This course

MHaton commented on lms/templates/instructor/edx_ace/enrollenrolled/email/body.html:
```diff
@@ -0,0 +1,30 @@
+            </h1>
+
+            <p style="color: rgba(0,0,0,.75);">
+                {% blocktrans %}You have been enrolled in {{ course_name }} at {{ site_name }} by a member of the course staff. The course should now appear on your {{ site_name }} dashboard.{% endblocktrans %}
```
The course should -> This course will

MHaton commented on lms/templates/instructor/edx_ace/enrollenrolled/email/body.txt:
```diff
@@ -0,0 +1,9 @@
+{% load i18n %}{% autoescape off %}{% blocktrans %}Dear {{ full_name }}{% endblocktrans %}
+
+{% blocktrans %}You have been enrolled in {{ course_name }} at {{ site_name }} by a member of the course staff. The course should now appear on your {{ site_name }} dashboard.{% endblocktrans %}
```
The course should -> This course will

nasthagiri commented on lms/templates/instructor/edx_ace/allowedenroll/email/body.html:
```diff
@@ -0,0 +1,62 @@
+                <br />
+            </p>
+
+            {% trans "Finish Your Regsitration" as button_cta_text %}
```
Typo: Registration

OmarIthawi replied:
Thanks @nasthagiri! Good catch :slightly_smiling_face: 

Fixed it.

OmarIthawi commented:
@MHaton Thanks! I've addressed all of your comments.

It'll take me a bit before I can get the tests passing.

OmarIthawi commented:
### AllowedUnenroll ace Message

<img src="https://user-images.githubusercontent.com/645156/45499522-43c02780-b785-11e8-8811-a57c24b91949.png" width="350" />

```
------- EMAIL -------
To: foo@gmail.com
From: 
Subject: You have been unenrolled from edX Demonstration Course
Body:
Dear Student,

You have been unenrolled from the course edX Demonstration Course by a member of the course staff. Please disregard the invitation previously sent.

----
This email was automatically sent from edx.devstack.lms:18000 to foo@gmail.com
-------  END  -------
```

OmarIthawi commented:
### EnrolledUnenroll ACE Message

<img src="https://user-images.githubusercontent.com/645156/45514253-8695f580-b7ad-11e8-8994-7c5c4545a65d.png" width="350" />

```
------- EMAIL -------
To: staff@example.com
From: 
Subject: You have been unenrolled from edX Demonstration Course
Body:
Dear Staff

You have been unenrolled from edX Demonstration Course at edx.devstack.lms:18000 by a member of the course staff. This course will no longer appear on your edx.devstack.lms:18000 dashboard.

Your other courses have not been affected.

----
This email was automatically sent from edx.devstack.lms:18000 to Staff
-------  END  -------
```

OmarIthawi commented:
# EnrollEnrolled ACE Message

<img src="https://user-images.githubusercontent.com/645156/45514860-40da2c80-b7af-11e8-8187-6bdd6374a0ef.png" width="350" />

```

------- EMAIL -------
To: staff@example.com
From: 
Subject: You have been enrolled in edX Demonstration Course
Body:
Dear Staff

You have been enrolled in edX Demonstration Course at edx.devstack.lms:18000 by a member of the course staff. This course will now appear on your edx.devstack.lms:18000 dashboard.

To start accessing course materials, please visit http://edx.devstack.lms:18000/courses/course-v1:edX+Demo+Demo_Course/

----
This email was automatically sent from edx.devstack.lms:18000 to Staff
-------  END  -------
```

OmarIthawi commented:
### AddBetaTester ACE Message

<img src="https://user-images.githubusercontent.com/645156/45515193-1dfc4800-b7b0-11e8-8525-11fb5d1d40a3.png" width="350" />

```
------- EMAIL -------
To: staff@example.com
From: 
Subject: You have been invited to a beta test for edX Demonstration Course
Body:
Dear Staff

You have been invited to be a beta tester for edX Demonstration Course at edx.devstack.lms:18000 by a member of the course staff.

To start accessing course materials, please visit http://edx.devstack.lms:18000/courses/course-v1:edX+Demo+Demo_Course/

----
This email was automatically sent from edx.devstack.lms:18000 to staff@example.com
-------  END  -------
```

OmarIthawi commented:
### RemoveBetaTester ACE Message

<img width="350" src="https://user-images.githubusercontent.com/645156/45515279-5e5bc600-b7b0-11e8-8750-fb6534c437c4.png" />

```
------- EMAIL -------
To: staff@example.com
From: 
Subject: You have been removed from a beta test for edX Demonstration Course
Body:
Dear Staff

You have been removed as a beta tester for edX Demonstration Course at edx.devstack.lms:18000 by a member of the course staff. This course will remain on your dashboard, but you will no longer be part of the beta testing group.

Your other courses have not been affected.

----
This email was automatically sent from edx.devstack.lms:18000 to staff@example.com
-------  END  -------
```

OmarIthawi commented:
### AccountCreationAndEnrollment ACE Message

This is activated by `ALLOW_AUTOMATED_SIGNUPS`, which arguably insecure!

<img src="https://user-images.githubusercontent.com/645156/45515663-754ee800-b7b1-11e8-8715-6a28b1964ee1.png" width="350" />

```
------- EMAIL -------
To: foobar@gmail.com
From: Your Platform Name Here
Subject: You have been enrolled in edX Demonstration Course
Body:
Welcome to edX Demonstration Course

To get started, please visit https://edx.devstack.lms:18000. The login information for your account follows.

email: foobar@gmail.com
password: zzzzzzz


It is recommended that you change your password.

Yours sincerely, The edX Demonstration Course Team
-------  END  -------
```

OmarIthawi commented:
Status: I've amended some emails and broke some tests along the way. Fixing those tests soon, probably tomorrow. After that this PR would be ready.

nedbat commented:
@MHaton I appreciate the fine-toothed copyediting pass.  Perhaps it's an American/British thing, but "Sincerely yours" is much more common on our side of the pond than "Yours sincerely."  Both sound kind of stuffy in this context anyway, but we are trying not to rewrite everything.

OmarIthawi commented:
Thanks @nedbat for pointing that out. Definitely the intention here is not to rewrite everything. There's a couple of the edits that I feel strongly about since some are actually typos like `You've been unenrolled in Course X`, where `in` should be `from`.

There are other fine-toothed copy-editing that I've made, but I'm a bit indifferent to reverting.

Apart from `Sincerely yours` do you think there's something that's better be reverted?

I'll be fixing the tests soon, so appreciate your feedback.

nedbat commented:
All the other edits sound great to me.

OmarIthawi commented:
@nedbat I've just fixed the tests (Unless bokchoy randomly fails), changed the mixin location as you've suggested, with few other minor modifications. This PR is :heavy_check_mark::heavy_check_mark: 

To track the changes better since your original review. An up to date diff since the commit `Hack: regenerate Esperanto lang so tests can pass` is available here:
 -  https://github.com/appsembler/edx-platform/compare/9e5725bb74540cff58f9fad7521a23e661b4acd5...omar/ace-enroll-email

I think you've reviewed most of those changes already, but the link is available anyway.

Thank you for the review!

OmarIthawi commented:
jenkins run bokchoy

OmarIthawi commented:
jenkins run bokchoy

bill-filler commented:
@OmarIthawi 
thanks for the heads up. I will have someone on the white-label team review to make sure the email message is themed correctly on WL sites.

OmarIthawi commented:
Thanks @bill-filler! Same goes for here, just in case it was buried in the notification stream: #18882

cc: @ziafazal

nedbat commented:
Resolve the conflicts in the translations, and I can merge it.

OmarIthawi commented:
Just resolved them, thanks @nedbat!

edx-webhook commented:
@OmarIthawi 🎉 Your pull request was merged!

Please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/edx-platform&entry.867055334=https://github.com/edx/edx-platform/pull/18475&entry.1484655318=https://github.com/OmarIthawi&entry.752974735=2018-06-27+17:14&entry.1917517419=2018-09-27+20:44&entry.2133058324=Yes) so we can improve your experience in the future.

OmarIthawi commented:
To all of the humans and bots who helped to push this PR forward, thank you!
